Store passwords as salted hashes.

haacked opened the pull request:
For existing users, this supports having a null salt. When users login, if
they don't have already have a salt, this will generate a salt for those users.
When users change passwords, they'll also get a salted version.

haacked committed:
Store passwords as salted hashes. Fixes #115

For existing users, this supports having a null salt. When users login, if
they don't have already have a salt, this will generate a salt for those users.
When users change passwords, they'll also get a salted version.

davidfowl commented on JabbR/Services/ChatService.cs:
```diff
@@ -9,10 +10,17 @@ namespace JabbR.Services
+        private readonly ICryptoService _crypto;
 
         public ChatService(IJabbrRepository repository)
+            : this(repository, new CryptoService())
```
Um ninject please :)

davidfowl commented on JabbR/Services/ChatService.cs:
```diff
@@ -54,10 +63,15 @@ namespace JabbR.Services
                 throw new InvalidOperationException(String.Format("Unable to claim '{0}'.", userName));
             }
+
+            if (String.IsNullOrEmpty(user.Salt))
```
Why this check?

haacked replied:
Because we only want to change your password to a salted password if you don't already have a salt. I could simply call ChangeUserPassword every time, but I thought that would be confusing. Obviously, this is confusing.

I could have an intermediate method simply named, `ChangePasswordToSaltedIfNotAlreadySalted` or something like that. :) Thoughts?

davidfowl replied:
Yea create a method that's verbose and as long a their is a unit test covering that scenario it's all good. A method name would help or maybe a comment. That code isn't very self documenting.

haacked replied:
Ok, I pushed a new commit that I think clarifies things. How's that?

davidfowl commented on JabbR/Services/ChatService.cs:
```diff
@@ -294,5 +314,17 @@ namespace JabbR.Services
+
+        private static string GenerateSalt()
+        {
+            byte[] data = new byte[0x10];
```
var

haacked replied:
Ack, this method doesn't belong here at all. It's a remnant. I'll remove it.

davidfowl commented on JabbR/Services/CryptoService.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public string CreateSalt()
+        {
+            byte[] data = new byte[0x10];
```
var

haacked committed:
Code cleanup based on suggestions from a code review

Removed an unused private method. Removed poor man's dependency injection. Etc.

haacked commented:
Ok, I made the requested changes.

haacked committed:
Added a method to clarify the code's intent

davidfowl commented:
Looks good. We'll see if you broke anything :). The staging site is here if you wanna see your changes semi-live:

http://jabbr-staging.apphb.com
